src/live_restore/live_restore_fs.c

```diff
@@ -804,57 +665,107 @@ __live_restore_fh_read(
+{
+    WTI_LIVE_RESTORE_FILE_HANDLE *lr_fh;
+    wt_off_t destination_size;
+
+    lr_fh = (WTI_LIVE_RESTORE_FILE_HANDLE *)fh;
+
+    WT_RET(lr_fh->destination.fh->fh_size(lr_fh->destination.fh, wt_session, &destination_size));
+    *sizep = destination_size;
+    return (0);
+}
+
+/*
+ * __live_restore_compute_read_end_bit --
+ *     Compute the last possible bit for hole filling.
+ */
+static int
+__live_restore_compute_read_end_bit(WT_SESSION_IMPL *session, WTI_LIVE_RESTORE_FILE_HANDLE *lr_fh,
+  wt_off_t buf_size, uint64_t first_clear_bit, uint64_t *end_bitp)
+{
+    wt_off_t read_start = WTI_BIT_TO_OFFSET(first_clear_bit);
+    wt_off_t file_size;
+    /*
+     * In theory we have truncated the destination file to be smaller than that of the source file.
+     * This would be better tracked with a variable on the lr_fh itself but for now we can work
+     * around it by reading the size of the destination file.
+     */
+    WT_RET(__live_restore_fh_size((WT_FILE_HANDLE *)lr_fh, (WT_SESSION *)session, &file_size));
+    file_size = WT_MIN(file_size, (wt_off_t)lr_fh->source_size);
+    wt_off_t largest_possible_read = WT_MIN(file_size, read_start + buf_size);
+    /* Sanity check. */
+    WT_ASSERT(session, lr_fh->destination.nbits == WTI_OFFSET_TO_BIT((wt_off_t)lr_fh->source_size));
+    /* Subtract 1 as the read end is served from the nbits - 1th bit.*/
+    uint64_t max_read_bit = WTI_OFFSET_TO_BIT(largest_possible_read) - 1;
+    uint64_t end_bit = first_clear_bit;
+    for (uint64_t current_bit = first_clear_bit + 1; current_bit <= max_read_bit; current_bit++) {
+        if (current_bit >= lr_fh->destination.nbits)
+            break;
+        if (__bit_test(lr_fh->destination.bitmap, current_bit))
+            break;
+        end_bit = current_bit;
+    }
+    *end_bitp = end_bit;
+    return (0);
+}
+
 /*
  * __live_restore_fill_hole --
  *     Fill a single hole in the destination file. If the hole list is empty indicate using the
  *     finished parameter. Must be called while holding the extent list write lock.
  */
 static int
-__live_restore_fill_hole(WT_FILE_HANDLE *fh, WT_SESSION *wt_session, char *buf,
-  WT_TIMER *start_timer, uint64_t *msg_count, bool *finishedp)
+__live_restore_fill_hole(WTI_LIVE_RESTORE_FILE_HANDLE *lr_fh, WT_SESSION *wt_session, char *buf,
+  wt_off_t buf_size, WT_TIMER *start_timer, uint64_t *msg_count, bool *finishedp)
 {
-    WTI_LIVE_RESTORE_FILE_HANDLE *lr_fh = (WTI_LIVE_RESTORE_FILE_HANDLE *)fh;
-    WTI_LIVE_RESTORE_HOLE_NODE *hole = lr_fh->destination.hole_list_head;
     WT_SESSION_IMPL *session = (WT_SESSION_IMPL *)wt_session;
 
-    WT_ASSERT(session, __wt_rwlock_islocked(session, &lr_fh->ext_lock));
-    if (hole == NULL) {
-        /* If there are no holes to fill we're done. */
+    WT_ASSERT(session, __wt_rwlock_islocked(session, &lr_fh->bitmap_lock));
+    uint64_t first_clear_bit;
+    if (__bit_ffc(lr_fh->destination.bitmap, lr_fh->destination.nbits, &first_clear_bit) == -1) {
```

Comment:
A comment would be useful about what bitmap state means completion would be helpful here. I had to think about what `bit_ffc` meant (find first cleared bit) and then translate that to mean that if all bits are set then it means we're done.

Reply:
8b9922c0b